File: clients/python_pypy/debug_interface.py
from codegame.client_message import ClientMessage
from debugging.debug_command import DebugCommand
from model.debug_state import DebugState
from model.debug_data import DebugData
from stream_wrapper import StreamWrapper
from model.vec2_double import Vec2Double
from stream_wrapper import StreamWrapper

class DebugInterface:
    def __init__(self, reader, writer):
        self.reader = reader
        self.writer = writer
        
    def add_circle(self, pos: Vec2Double, radius: float):
        self.add(DebugData.Circle(pos, radius))

    def add(self, debug_data: DebugData):
        self.send(DebugCommand.Add(debug_data))
    
    def clear(self):
        self.send(DebugCommand.Clear())
    
    def set_auto_flush(self, enable: bool):
        self.send(DebugCommand.SetAutoFlush(enable))
    
    def flush(self):
        self.send(DebugCommand.Flush())

    def send(self, command: DebugCommand):
        ClientMessage.DebugMessage(command).write_to(self.writer)
        self.writer.flush()

    def get_state(self) -> DebugState:
        ClientMessage.RequestDebugState().write_to(self.writer)
        self.writer.flush()
        return DebugState.read_from(self.reader)

File: clients/cpp_clang/DebugInterface.cpp
#include "DebugInterface.hpp"
#include "codegame/ClientMessage.hpp"

DebugInterface::DebugInterface(TcpStream* stream): stream(stream) {}

void DebugInterface::addCircle(model::Vec2Double pos, double radius, debugging::Color color)
{
    add(debugging::DebugData(debugging::Circle(pos, radius, color)));
}

void DebugInterface::addLine(model::Vec2Double point1, model::Vec2Double point2, double width, debugging::Color color)
{
    add(debugging::DebugData(debugging::Line(point1, point2, width, color)));
}

void DebugInterface::addRect(model::Vec2Double corner1, model::Vec2Double corner2, debugging::Color color)
{
    add(debugging::DebugData(debugging::Rect(corner1, corner2, color)));
}

void DebugInterface::addText(std::string text, model::Vec2Double pos, double size, double align, debugging::Color color)
{
    add(debugging::DebugData(debugging::Text(text, pos, size, align, color)));
}

void DebugInterface::add(debugging::DebugData debugData)
{
    send(debugging::DebugCommand(debugging::Add(debugData)));
}

void DebugInterface::clear()
{
    send(debugging::DebugCommand(debugging::Clear()));
}

void DebugInterface::setAutoFlush(bool enable)
{
    send(debugging::DebugCommand(debugging::SetAutoFlush(enable)));
}

void DebugInterface::flush()
{
    send(debugging::DebugCommand(debugging::Flush()));
}

void DebugInterface::send(debugging::DebugCommand command)
{
    codegame::ClientMessage message = codegame::DebugMessage(command);
    codegame::writeClientMessage(message, *stream);
    stream->flush();
}

debugging::DebugState DebugInterface::getState()
{
    codegame::ClientMessage message = codegame::RequestDebugState();
    codegame::writeClientMessage(message, *stream);
    stream->flush();
    return debugging::DebugState::readFrom(*stream);
}